### ISSUE TITLE
Actually actionable list of batching rules to write

### ISSUE THREAD

**Issue by zou3519** [pytorch/functorch#240]
For each of the items here, we should make sure all compositions have a batching rule. All of these items should be actionable (in that it is possible to write a batching rule and we are not blocked on functionalization, which is coming soon).

Note: you may need to write an OpInfo for the operator if it doesn't exist already or wait for one to be added. A lot of folks are adding OpInfos right now, so if the OpInfo doesn't exist please ask first to see if someone is working on it.

Parcel 1: top nn.functional.* and top torch.* foo
- [ ] torch.nn.functional.interpolate (this involves adding batching rules for adaptive_avg_pool1d, but we might as well do `adaptive_avg_pool{1, 2, 3}d` as well as their backward variants while we're at it)
- [ ] nn.functional.pad. The batching rule seems to have a bug; the goal is to debug it.
- [ ] nn.functional.unfold. Involves writing a batching rule for im2col, im2col_backward. I don't know how difficult these are.
- [ ] nn.functional.grid_sample. Involves writing a batching rule for the backward operator.
- [ ] torch.pow. The backward needs batching rule for logical_and ; use this as an opportunity to write the batching rules for `logical_{and, or, xor}` if those don't exist yet.

**Comment by vfdev-5** [comment]
@zou3519 `nn.functional.pad` with circular option requires to do a copy: `out[..., out_d0:out_d1] = input[..., in_d0:in_d1]` and thus there is the following error is raised:
```
>           out[..., out_d0:out_d1] = input[..., in_d0:in_d1]                                                                                                                               
E           RuntimeError: vmap: aten::copy_(self, *extra_args) is not possible because there exists a Tensor `other` in extra_args that has more elements than `self`. This happened due to `other` being vmapped over but `self` not being vmapped over at level 2. Please try to use out-of-place operators instead of aten::copy_. If said operator is being called inside the PyTorch framework, please file a bug report instead. 
```
Can we do something with that ?

**Comment by zou3519** [comment]
Aha. Nope, we can't do anything about that until functionalization is in, good catch.

**Comment by Padarn** [comment]
Hi @zou3519, the "forward pass only" ops above means that the vjp and related operators require the functionalization too?

**Comment by zou3519** [comment]
> Hi @zou3519, the "forward pass only" ops above means that the vjp and related operators require the functionalization too?

Yes, "forward pass only" means we should only try to get the `vmap` tests passing and none of the vjp/grad/compositions of {vjp, grad} tests.

**Comment by vfdev-5** [comment]
@kshitij12345 on which tasks from Parcel 2 you are working on ?

**Comment by kshitij12345** [comment]
@vfdev-5 I think I'll be picking `diagonal_scatter` next. Go ahead with `_cdist_forward, _cdist_backward`

**Comment by vfdev-5** [comment]
I'll take `torch.addr` next

**Comment by vfdev-5** [comment]
Let's hold on on svd batch rule as there is ongoing refactoring which may fix CPU/CUDA discrepancy test issue: https://github.com/pytorch/pytorch/pull/69827

**Comment by vfdev-5** [comment]
To finish this issue, it remains to finalize parcel 2:
- https://github.com/pytorch/functorch/pull/322
and in parcel 4:
- index_add was already done : https://github.com/pytorch/pytorch/pull/67329
- index_select is done but not merged : https://github.com/pytorch/pytorch/pull/67329
- index_fill : todo
- masked_fill : todo
- masked_scatter : todo

**Comment by zou3519** [comment]
There's always more batching rules to write, I'll put up a new issue for them later :)

**Comment by Lezcano** [comment]
Note that `_lu_with_info` is not a thing any more. Now we have `linalg_lu_factor` and `linalg_lu_factor_ex` cf. https://github.com/pytorch/pytorch/pull/66933

**Comment by Lezcano** [comment]
It will indeed. And that's a good reminder for me to put up a PR doing so :D

**Comment by vfdev-5** [comment]
@zou3519 can we update description list with with was done. I think we can remove Parcel 4 from here and create new issue for that if needed. What remains here is to sync and merge householder product PR (#322), cc @kshitij12345 .

**Comment by Lezcano** [comment]
Fwiw, following up on the point above on deprecating `torch.lu`:
https://github.com/pytorch/pytorch/pull/73804
https://github.com/pytorch/pytorch/pull/73806

**Comment by zou3519** [comment]
> @zou3519 can we update description list with with was done. I think we can remove Parcel 4 from here and create new issue for that if needed. What remains here is to sync and merge householder product PR (#322)

Yes I'll create another issue soon
